단축키 적용 및 구간 선택기능 구현

Songwonseok opened the pull request:
## 📕 제목

단축키 적용 및 구간 선택기능 구현
관련이슈 #71 

## 📗 작업 내용

> 구현 내용 및 작업 했던 내역

- [x] 단축키 이벤트 적용
- [x] 구간 선택기능 구현
- [x] 선택 조건에 따라 EditTools 컴포넌트 렌더링 변경

![녹화_2020_12_03_12_35_32_447](https://user-images.githubusercontent.com/7006837/100960752-44128c00-3564-11eb-9111-1f330905c367.gif)

## 📘 PR 특이 사항

> PR을 볼 때 주의깊게 봐야하거나 말하고 싶은 점

- 단축키
  - App 컴포넌트에서 window에 keydown, keyup 이벤트 등록했습니다.
  - keyDown 이벤트에 조건문 안에 단축키 관련 함수들을 넣어주면 됩니다.

- 구간 선택
  - 구간을 선택하면 focusList에 구간에 대한 정보(TrackSection, HTMLElement)가 저장됩니다.
    - focus된 구간에 command를 적용시키기 위해서 TrackSection 정보가 저장됩니다.
    - 선택된 구간에 style 적용시키기 위해서 HTMLElement가 저장됩니다. 
  - 구간 선택 관련해서 controller의 toggleFocus() 내부에서 조건에 따라 적용됩니다.
  - ctrl키 눌렀는지에 대한 조건에 따라 focusList의 값이 변화
    - ctrl키 누르고 구간을 선택한 경우
      - 여러 구간 선택 가능
      - 이미 선택된 구간을 누르면 취소
    - ctrl키 누르지 않고 구간을 선택한 경우
      - 기존에 focusList를 reset하고 다시 선택


- EditTools 렌더링 조건
  - cursorMode, focusList, clipBoard, redo/undo list 에 따라 아이콘 버튼
    활성화
  - cursorMode는 현재 마우스가 선택하기 or 나누기 모드인지 확인
  - focusList는 현재 몇개의 구간을 선택했는 지 확인
    - 0개면 복사, 잘라내기, 삭제 모두 비활성화
    - 1개면 복사, 잘라내기, 삭제 아이콘 모두 활성화
    - 2개 이상이면 삭제 아이콘만 활성화
  - clipBoard는 현재 복사하거나 잘라내기한 객체가 clipBoard에 저장되어
    있는지 확인
    - 객체가 존재한다면 붙여넣기 아이콘 활성화
    - 현재 focusList의 길이가 2이상이면 비활성화
  - undoList, redoList의 길이가 1이상이면 각 undo/redo 아이콘이
    활성화됩니다.

Songwonseok committed:
- border가 겹쳐서 나와서 audi-effect-controller의 border를 제거

Songwonseok committed:
feat: [FE] 단축키 이벤트 구현

- App에서 windows addEventListener로 단축키 이벤트 등록
  - App에서 this.addEventListener는 단축키가 안먹더라구요...
- store의 ctrlIsPressed로 ctrl키가 눌려있는지 확인

Songwonseok committed:
feat: [FE] store에 ctrlIsPressed 추가

- ctrl 키가 눌렸는지 확인할 수 있는 ctrlIsPressed 변수 추가

Songwonseok committed:
refactor: [FE] KeyBoard enum 모듈화, 단축키 추가

- 건너뛰기, 재생/정지 때 필요한 LEFT,RIGHT, SPACE 단축키 설정
- KeyBoard enum을 types에 모듈화

Songwonseok committed:
chore: [FE] @command alias 추가

- redo/undo를 위한 Command 관련 디렉토리 alias 적용

Songwonseok committed:
feat: [FE] auto Increment 적용 및 focusList 관련내용 추가

- 구간 선택 관련해서 controller의 toggleFocus()에서 실행됩니다.
- ctrl키 눌렀는지에 대한 조건에 따라 focusList의 값이 변화
  - ctrl키 누르고 구간을 선택한 경우
    - 여러 구간 선택 가능
    - 이미 선택된 구간을 누르면 취소
  - ctrl키 누르지 않고 구간을 선택한 경우
    - 기존에 focusList를 reset하고 다시 선택
- focusList에는 TrackSection과 그 section의 canvas
      HTMLElement이 저장됩니다.
    - 선택된 구간에 style 적용시키기 위해서 저장

Songwonseok committed:
feat: [FE] 구간 선택 AudioTrackSection에 적용

- 구간을 클릭했을 때 toggleFocus() 실행하도록 구현
- 선택된 구간에는 focused-section 클래스를 적용해서 style 변경했습니다.
  - 나중에 더 예쁜걸로 바꿀게요...

Songwonseok committed:
feat: [FE] 구간 선택기능을 위한 이벤트키 등록

- 구간 선택을 위한 AUDIO_TRACK_SECTION_CLICK 등록
- 다른 범위를 클랙했을 때 구간 선택 초기화 되기 위한 FOCUS_RESET_CLICK
  등록

Songwonseok committed:
feat: [FE] 구간선택 초기화 이벤트를 위한 범위 설정

- ctrl키를 누르지 않은 상태에서 audi-main-audio-track-container 영역을
  클릭하면 기존에 선택된 focus가 초기화됩니다.
- Main과 AudioTrack 둘다 이벤트를 적용했을 때 부자연스럽지 않아서 두 곳
  모두 이벤트를 적용했습니다.

Songwonseok committed:
feat: [FE] EditTool 컴포넌트 아이콘 활성화 조건 추가

- cursorMode, focusList, clipBoard, redo/undo list 에 따라 아이콘 버튼
  활성화
- cursorMode는 현재 마우스가 선택하기 or 나누기 모드인지 확인
- focusList는 현재 몇개의 구간을 선택했는 지 확인
  - 0개면 복사, 잘라내기, 삭제 모두 비활성화
  - 1개면 복사, 잘라내기, 삭제 아이콘 모두 활성화
  - 2개 이상이면 삭제 아이콘만 활성화
- clipBoard는 현재 복사하거나 잘라내기한 객체가 clipBoard에 저장되어
  있는지 확인
  - 객체가 존재한다면 붙여넣기 아이콘 활성화
- undoList, redoList의 길이가 1이상이면 각 undo/redo 아이콘이
  활성화됩니다.

Songwonseok committed:
refactor: [FE] EditTools clipBoardState() 조건 변경

- 2개 이상 선택되있거나 clipBoard에 객체가 없을 때 붙여넣기 비활성화

Jeongeun-Choi commented on frontend/src/components/App/App.ts:
```diff
@@ -32,10 +33,61 @@ import { EventDataType, eventTypes, EventTargetDataType } from '@types';
 
     initEvent(): void {
       this.eventsForListener.forEach((eventName) => this.addEventListener(eventName, this.eventListenerForRegistrant.bind(this)));
+      addEventListener('keydown', this.KeyDownListener)
+      addEventListener('keyup', this.ctrlKeyUpListener)
+    }
```
window.addEventListener 라고 표기해주는게 명시적이고 좋을 것 같아요!😊

Songwonseok replied:
window 표기하는게 더 좋겠네요 수정해서 커밋하겠습니다

wooojini commented on frontend/src/common/types/keyType.ts:
```diff
@@ -0,0 +1,14 @@
+enum KeyBoard {
+  Z = 90,
+  Y = 89,
+  X = 88,
```
오 키보드 값인가 보군요!! :-)

wooojini commented on frontend/src/components/App/App.ts:
```diff
@@ -32,10 +33,61 @@ import { EventDataType, eventTypes, EventTargetDataType } from '@types';
+        Controller.setCtrlIsPressed(true);
+      }
+
+      if (e.which === KeyBoard.C && !isCtrl) {
```
keyDown 이벤트 리스너의 처리는 처음보네요!!! 굉장히 오랜만에 보는 분기 연속처리네요... 조합들을 생각하느라 고생하셨을 거 같습니다! 잘 이용해볼게요! 💯

Songwonseok replied:
원래는 각 단축키별로 다 keyDown을 먹였었는데 하나에만 해도 다 되더라구요

wooojini commented on frontend/src/controllers/controller.ts:
```diff
@@ -64,6 +59,81 @@ const addTrackSection = (trackId: number, trackSection: TrackSection): void => {
     store.setTrackSection(trackId, trackSection);
 }
 
+const getCtrlIsPressed = (): boolean => {
```
store에서 get하는 함수들은 getState를 활용하는 것은 어떨까요??? getter함수들은 단순히 store의 데이터를 그대로 전달해주기만 하는데, store의 유지하는 데이터가 늘어날 때마다 getter함수가 늘어날 것 같습니다!!
```
  //아래와 같이 controller에서 getState()로 가져와서 사용하는 방식
  const {  ctrlIsPressed } = store.getState();
  ....
```

Songwonseok replied:
아 맞네요...ㄷㄷ getState()가 있었네요
수정하겠습니다!

wooojini replied:
제가 잘못언급했습니다 ㅠㅠㅠ 무시해주세용!

wooojini commented on frontend/src/store/store.ts:
```diff
@@ -61,46 +67,77 @@ const store = new (class Store{
-        const track = trackList[trackId];
         const { trackSectionList } = track;
-        newTrackSection.id = trackSectionList.length;
+        newTrackSection.id = this.state.sectionIndex++;
```
AutoIncrement하는 부분이군요 :-)

wooojini reviewed (approved):
고생하셨습니다 :-) 단축키라니 이번주는 하루하루 여러분들의 작업에 놀라고 있습니다 ㅎㅎ

Jeongeun-Choi commented on frontend/src/store/store.ts:
```diff
@@ -61,46 +67,77 @@ const store = new (class Store{
         });
     }
+
+    setCtrlIsPressed(isPressed: boolean): void {
+        this.state.ctrlIsPressed = isPressed;
+    }
+
+    addFocus(newFocusInfo: FocusInfo): void {
+        const { focusList } = this.state;
+        const newfocusList = focusList.concat(newFocusInfo);
+
+        this.state = { ...this.state, focusList: newfocusList };
+        storeChannel.publish(StoreChannelType.EDIT_TOOLS_CHANNEL, '');
+    }
+
+    removeFocus(removeIndex: number) {
```
포커스와 단축키 이벤트가 이렇게 동작되는군요! 멋지십니다🤩

Jeongeun-Choi reviewed (approved):
가슴이 웅장해집니다... 수고하셨습니다😄

Songwonseok committed:
refactor: [FE] 단축키 이벤트 window 표기

- window를 표기하는게 명시적으로 좋을 것 같아서 수정
- addEventListener -> widow.addEventListener로 수정했습니다.

pieisland reviewed (approved):
오 단축키는 이렇게 해서 작동되는 거였군요. 좋네요~
고생 많으셨습니다!

wooojini committed:
Merge branch 'develop' into feature/mouse-section-focus
